File: python/run-length-encoding/run_length_encoding.py
import re
from functools import reduce

def decode(string):
    quantities = re.findall('(\d+)?(\w|\s)', string)

    return reduce(__decode, quantities, '')


def __decode(result, quantity):
    count = int(quantity[0]) if quantity[0] else 1
    result += count * quantity[1]
    return result


def encode(string):
    return __encode(string, '')


def __encode(string, result):
    if len(string) == 0: return result

    match = re.match('{}+'.format(string[0]), string)
    count = len(match.group())

    if count > 1: result += str(count)
    result += match.group()[0]

    remaining_string = string[match.end():]

    return __encode(remaining_string, result)

File: ruby/run-length-encoding/run_length_encoding.rb
require 'byebug'
class RunLengthEncoding
  class << self
    def encode(input)
      input = input.dup

      _encode(input, '')
    end

    def decode(input)
      scan(/(\d+)?(\w|\s)/).inject("") do |decoded_string, (count, char)|
        count = count ? count.to_i : 1
        decoded_string << char * count
      end
    end

    private

      def _encode(str, result)
        return result if str.empty?

        str[/#{str[0]}+/]

        # $& is the match and $` is the string after the match (remainder)
        result << $&.size.to_s if $&.size > 1
        result << $&[0]

        _encode $', result
      end
  end
end

class BookKeeping
  VERSION = 3
end
